apps/els_core/src/els_text.erl

```diff
@@ -52,6 +59,61 @@ last_token(Text) ->
     Tokens -> lists:last(Tokens)
   end.
 
+-spec apply_edits(text(), [edit()]) -> text().
+apply_edits(Text, []) ->
+  Text;
+apply_edits(Text, Edits) when is_binary(Text) ->
+  lines_to_bin(apply_edits(bin_to_lines(Text), 1, 0, lists:sort(Edits))).
```

Comment:
AFAIK you can't sort edits - this can result in different content. They have to be applied in the order they are received. VS Code most of the time will sort them, so they are easier to apply, but this is not guaranteed, especially from other editors.

See this note in [documentation](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_didChange):
```
	/**
	 * The actual content changes. The content changes describe single state
	 * changes to the document. So if there are two content changes c1 (at
	 * array index 0) and c2 (at array index 1) for a document in state S then
	 * c1 moves the document from S to S' and c2 from S' to S''. So c1 is
	 * computed on the state S and c2 is computed on the state S'.
	 *
	 * To mirror the content of a document using change events use the following
	 * approach:
	 * - start with the same initial content
	 * - apply the 'textDocument/didChange' notifications in the order you
	 *   receive them.
	 * - apply the `TextDocumentContentChangeEvent`s in a single notification
	 *   in the order you receive them.
	 */
```

Reply:
Thanks for pointing this out!
lsp-mode in Emacs only ever sent edits one at a time, so I never actually tested this in a "real world" scenario.
It makes the implementation a lot simpler actually :+1: